lib/AST/ASTContext.cpp

```diff
@@ -8407,6 +8413,48 @@ bool ASTContext::isEqualIgnoringChecked(QualType T1, QualType T2) const {
   }
 }
 
+// For the Checked C extension, compute whether a type is allowed to be an
+// argument or return type for a no-prototype function.   This computes the
+// set of allowed types described in Section 5.5 of the Checked C
+// specification.
+bool ASTContext::isNotAllowedForNoProtoTypeFunction(QualType QT) const {
+  if (const PointerType *PT = QT->getAs<PointerType>()) {
+    if (PT->isChecked())
+      return true;
+    if (PT->isFunctionPointerType())
+      return isNotAllowedForNoProtoTypeFunction(PT->getPointeeType());
+    // Unchecked pointer types are allowed
+    return false;
+  } else if (QT->isCheckedArrayType())
```

Comment:
Yes, you can have an unchecked array of function pointers where the function pointers have params or return types that are checked.    They are treated the same as an unchecked pointer type (allowed for no prototype functions and allowed for members of structures passed to no prototype functions.  It is clear in the code that the member isn't a checked.   There's an implicit conversion for uses of the array type to the unchecked pointer type.

You are right to get nervous - I was hesitant about this method about that about it before creating it.  Right now, the definitions are subtly different, so creating methods with generic functionality seemed more likely to cause confusion down the road than creating a very specific function that ties directly to text in the Checked C spec.

It's easy to diff the code later and see how many different concepts we ended up with and figure out whether we can/should unify the functionality.